File: src/main/java/me/imoltres/bbu/utils/config/MainConfig.java
package me.imoltres.bbu.utils.config;

import me.imoltres.bbu.BBU;

import java.util.ArrayList;
import java.util.LinkedHashMap;

/**
 * Configuration properties imported from config.yml
 */
public class MainConfig<T> extends ConfigGetter<T> {

    public static int BORDER_SIZE = new MainConfig<Integer>("border.size").get();
    public static ArrayList<LinkedHashMap<String, Integer>> BORDER_PHASES = new MainConfig<ArrayList<LinkedHashMap<String, Integer>>>("border.shrink-phases").get();

    public static boolean DEATHMATCH_ENABLED = new MainConfig<Boolean>("deathmatch.enabled").get();
    public static int DEATHMATCH_TIME = new MainConfig<Integer>("deathmatch.time").get();

    public static boolean SPECTATE_AFTER_DEATH = new MainConfig<Boolean>("spectate-after-death").get();

    public static boolean FRIENDLY_FIRE = new MainConfig<Boolean>("friendly-fire").get();

    public static String LOBBY_SPAWN = new MainConfig<String>("lobby-spawn").get();
    public static String DEATHMATCH_SPAWN = new MainConfig<String>("deathmatch.spawn").get();

    MainConfig(String path) {
        super(BBU.getInstance().getMainConfig(), path);
    }

    public static void reload() {
        BBU.getInstance().getMainConfig().reloadConfig();

        //i really wish there was a better way to do this
        BORDER_SIZE = new MainConfig<Integer>("border.size").get();
        BORDER_PHASES = new MainConfig<ArrayList<LinkedHashMap<String, Integer>>>("border.shrink-phases").get();

        DEATHMATCH_ENABLED = new MainConfig<Boolean>("deathmatch.enabled").get();
        DEATHMATCH_TIME = new MainConfig<Integer>("deathmatch.time").get();

        SPECTATE_AFTER_DEATH = new MainConfig<Boolean>("spectate-after-death").get();

        FRIENDLY_FIRE = new MainConfig<Boolean>("friendly-fire").get();

        LOBBY_SPAWN = new MainConfig<String>("lobby-spawn").get();
        DEATHMATCH_SPAWN = new MainConfig<String>("deathmatch.spawn").get();
    }

}

File: src/main/java/me/imoltres/bbu/utils/config/MainConfig.kt
package me.imoltres.bbu.utils.config

import me.imoltres.bbu.BBU
import me.imoltres.bbu.game.ShrinkPhase

class MainConfig<T>(path: String) : ConfigGetter<T>(BBU.getInstance().mainConfig, path) {

    companion object {
        var borderSize = 0
        var borderPhases = listOf<ShrinkPhase>()

        var deathmatchEnabled = false
        var deathmatchTime = 0
        var deathmatchSpawn = ""
        var deathmatchTeamLocations = listOf<String>()

        var spectateAfterDeath = false
        var friendlyFire = false

        var lobbySpawn = ""
        var lobbyYMin = 0

        var beaconYMin = 0
        var beaconYMax = 0

        init {
            load()
        }

        private fun <T> cfg(path: String): T = MainConfig<T>(path).get()

        private fun load() {
            borderSize = cfg("border.size")
            borderPhases = cfg<List<LinkedHashMap<String, Int>>>("border.shrink-phases")
                .mapTo(mutableListOf()) { ShrinkPhase(it["size"]!!, it["time"]!!, it["length"]!!) }
            deathmatchEnabled = cfg("deathmatch.enabled")
            deathmatchTime = cfg("deathmatch.time")
            deathmatchSpawn = cfg("deathmatch.spawn")
            deathmatchTeamLocations = cfg("deathmatch.team-locations")
            spectateAfterDeath = cfg("spectate-after-death")
            friendlyFire = cfg("friendly-fire")
            lobbySpawn = cfg("lobby.spawn")
            lobbyYMin = cfg("lobby.y-min")
            beaconYMin = cfg("beacon.y-min")
            beaconYMax = cfg("beacon.y-max")
        }

        fun reload() {
            BBU.getInstance().mainConfig.reloadConfig()
            load()
        }
    }
}